Add cgroup mount in the recommended config

hqhq opened the pull request:
Otherwise we have no way to know our resource limit in the runtime.

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

avagin commented:
If you want use them only to know resource limits, it's better to mount them read-only, isn't it?

hqhq commented:
Yes, it needs to be read-only, but currently libcontainer won't take cgroup mount options, see https://github.com/opencontainers/runc/blob/master/libcontainer/rootfs_linux.go#L172

So it doesn't matter what mount option we set in spec, I don't see the needs we customize cgroup mount option, if we need, we can change that logic in the future.

avagin commented:
Actually I found that cgroups are mounted as read-only without any special options.
https://github.com/opencontainers/runc/blob/master/libcontainer/rootfs_linux.go#L187

Can we remove the options field if it's ignored?

crosbymichael commented:
@hqhq @avagin we should probably change that hard coded RO to respect the flags

mrunalp commented:
#110 created to take out the hard-coded RO

hqhq commented:
@crosbymichael @mrunalp @avagin 

I have updated PR. I think we should keep adding MS_RDONLY option in libcontainer, because that's a security issue, if users can mount cgroup as rw, the container can change resources which will affect the host.

The problem before is it didn't take any flags from config, now I added that. PTAL.

mrunalp commented:
@hqhq Yeah, probably a good idea to set it readonly, but some people have brought up use cases to make it non readonly. So, I guess it makes sense to allow it to be configurable and we can recommend readonly.

hqhq commented:
@mrunalp Yes, we do have use cases for cgroup to be non read-only, as https://github.com/docker/docker/pull/14575 said, dind would need cgroup to be rw, I updated commits to remove read-only restriction, I think we can add restriction on Docker side, like only allow cgroup mount to be rw in `--privileged` mode.
@crosbymichael PTAL.

LK4D4 commented:
@mrunalp @hqhq Can you enlighten me about relation of this to #110?

hqhq commented:
@LK4D4 Yeah that's kind of duplicate, @mrunalp tried to help but I think we can handle it in this PR.
And #110 still missed a point that we need to take flags from user configs.
Which one do you think we should take? I'm ok with both as long as #110 can be changed as I suggested.

LK4D4 commented:
I think here we have more checks. And also @mrunalp as maintainer can merge this :)

LK4D4 commented on libcontainer/rootfs_linux.go:
```diff
@@ -174,6 +175,10 @@ func mountToRootfs(m *configs.Mount, rootfs, mountLabel string) error {
 		}
+		flags := m.Flags
+		if flags&syscall.MS_RDONLY != 0 {
+			logrus.Warn("Cgroups are not mounted as read-only, container might affect host's resources.")
```
Is this happens inside container?

hqhq replied:
I thinks so, it happens in child process. But we'll still get this warning, right?

LK4D4 replied:
No, I think we don't, we just break our output. I think I'm not ready for introducing logging to child, maybe only if there will be `logrus.SetOutput(ioutil.Discard)` by default.

mrunalp replied:
Yes, we shouldn't log anything here.

hqhq replied:
@LK4D4 @mrunalp Updated, what do you think now?

mrunalp commented on spec.go:
```diff
@@ -223,6 +229,9 @@ func createLibcontainerMount(cwd string, m specs.Mount) *configs.Mount {
 			source = filepath.Join(cwd, m.Source)
 		}
 	}
+	if m.Type == "cgroup" && flags&syscall.MS_RDONLY == 0 {
```
I think we can take this out. runc is low level enough for people to know what they are doing. Also, as we show ro in the recommendation, I think it should suffice.

hqhq replied:
OK, let Docker concern about this, I can live with that :) 
updated and squished the commits, PTAL.

LK4D4 commented on spec.go:
```diff
@@ -83,6 +83,12 @@ var specCommand = cli.Command{
+						Type:        "cgroup",
+						Source:      "cgroup",
+						Destination: "/sys/fs/cgroup",
+						Options:     "nosuid,noexec,nodev,relatime,ro",
```
I have question, shouldn't we mount `/sys/fs/cgroup` as `ro` too? because now we mount only subsystems as `ro`.

LK4D4 replied:
Seems like it should be mounted as `ro` always.

hqhq replied:
If `/sys/fs/cgroup` mounted as `ro`, we can't create subsystem dirs, I know it seems a bit strange since it looks different from what it shows.

LK4D4 replied:
We can remount it later. So it will be exactly like on host.

LK4D4 commented on README.md:
```diff
@@ -111,6 +111,12 @@ user named `daemon` defined within that file-system.
             "options": "nosuid,noexec,nodev"
+        },
+        {
+            "type": "cgroup",
```
I don't know, this looks like  pretty confusing, because actually we ignoring "source".

hqhq replied:
Yeah, that's because we use bind mount, and I don't think that part of details should be exposed to users.

LK4D4 commented:
Okay, at least this fixes one more problem. LGTM

mrunalp commented:
LGTM
